syft/source/source.go

```diff
@@ -26,28 +29,38 @@ type Source struct {
 	directoryResolver *directoryResolver
 	path              string
 	mutex             *sync.Mutex
+	Exclusions        []string
 }
 
 type sourceDetector func(string) (image.Source, string, error)
 
 // New produces a Source based on userInput like dir: or image:tag
-func New(userInput string, registryOptions *image.RegistryOptions) (*Source, func(), error) {
+func New(userInput string, registryOptions *image.RegistryOptions, exclusions ...string) (*Source, func(), error) {
 	fs := afero.NewOsFs()
 	parsedScheme, imageSource, location, err := detectScheme(fs, image.DetectSource, userInput)
 	if err != nil {
 		return &Source{}, func() {}, fmt.Errorf("unable to parse input=%q: %w", userInput, err)
 	}
 
+	source := &Source{}
+	fn := func() {}
+
 	switch parsedScheme {
 	case FileScheme:
-		return generateFileSource(fs, location)
+		source, fn, err = generateFileSource(fs, location)
 	case DirectoryScheme:
-		return generateDirectorySource(fs, location)
+		source, fn, err = generateDirectorySource(fs, location)
 	case ImageScheme:
-		return generateImageSource(location, userInput, imageSource, registryOptions)
+		source, fn, err = generateImageSource(location, userInput, imageSource, registryOptions)
+	default:
+		err = fmt.Errorf("unable to process input for scanning: '%s'", userInput)
+	}
+
+	if err == nil {
+		source.Exclusions = exclusions
```

Comment:
Does it make sense to make exclusions compacted into where we generate the source that way we don't have to modify it after it has been returned from the specific constructors?

This would also allow us to keep the same branching logic from before.

Not a hard ask or anything just a thought since I think this might have to be refactored when taking into account windows paths. Each generate function will have to do some specific mutation to the exclusions to make it match with the source being generated in some instances.

Reply:
The problem is I need this data when `FileResolver(...)` is called, so it doesn't really do a lot of good to pass it down and the only caller is in the `cmd` package, which doesn't seem like the right place to set this (and makes it harder to test)